types/src/ledger_info.rs

```diff
@@ -475,66 +498,43 @@ impl LedgerInfoWithUnverifiedSignatures {
 
     fn try_aggregate(
         &mut self,
-        epoch_state: &EpochState,
+        verifier: &ValidatorVerifier,
     ) -> Result<AggregateSignature, VerifyError> {
-        self.check_voting_power(&epoch_state.verifier, true)?;
+        self.check_voting_power(verifier, true)?;
 
         let all_signatures = self
-            .verified_signatures
-            .signatures_iter()
-            .chain(self.unverified_signatures.signatures_iter());
-
-        epoch_state.verifier.aggregate_signatures(all_signatures)
-    }
-
-    /// Merge unverified signatures into verified signatures if they are valid.
-    fn merge_signatures(&mut self, verifier: &ValidatorVerifier, need_verify: bool) {
-        let unverified_signatures =
-            mem::replace(&mut self.unverified_signatures, PartialSignatures::empty()).unpack();
-        let valid_signatures: Vec<_> = unverified_signatures
-            .into_par_iter()
-            .flat_map(|(account_address, signature)| {
-                if !need_verify
-                    || verifier
-                        .verify(account_address, self.ledger_info(), &signature)
-                        .is_ok()
-                {
-                    Some((account_address, signature))
-                } else {
-                    verifier.add_pessimistic_verify_set(account_address);
-                    None
-                }
-            })
-            .collect();
-        for (account_address, signature) in valid_signatures {
-            self.verified_signatures
-                .add_signature(account_address, signature);
-        }
+            .signatures
+            .iter()
+            .map(|(voter, sig)| (voter, sig.signature()));
+        verifier.aggregate_signatures(all_signatures)
+    }
+
+    fn filter_invalid_signatures(&mut self, verifier: &ValidatorVerifier, need_verify: bool) {
+        let signatures = mem::take(&mut self.signatures);
+        self.signatures =
+            verifier.filter_invalid_signatures(self.ledger_info(), signatures, need_verify);
     }
 
     /// Try to aggregate all signatures if the voting power is enough. If the aggregated signature is
     /// valid, return the LedgerInfoWithSignatures. Also merge valid unverified signatures into verified.
     pub fn aggregate_and_verify(
         &mut self,
-        epoch_state: Arc<EpochState>,
+        verifier: &ValidatorVerifier,
     ) -> Result<LedgerInfoWithSignatures, VerifyError> {
-        let aggregated_sig = self.try_aggregate(&epoch_state)?;
+        let aggregated_sig = self.try_aggregate(verifier)?;
 
-        match epoch_state
-            .verifier
-            .verify_multi_signatures(self.ledger_info(), &aggregated_sig)
-        {
+        match verifier.verify_multi_signatures(self.ledger_info(), &aggregated_sig) {
             Ok(_) => {
-                self.merge_signatures(&epoch_state.verifier, false);
+                self.filter_invalid_signatures(verifier, false);
```

Comment:
in some sense we don't really need this since the aggregation already succeeds and returns?

Reply:
Yeah, we don't. But it still makes the data structure complete, and makes it easy in test cases to check that after aggregate_and_verify function is called, we have x number of verified signatures.

Reply:
Alin was telling me that even if the aggregation succeeds, it doesn't necessarily mean all individual signatures are valid, some validators can collude to produce two invalid signatures that can be aggregated to be valid. in that sense, for the sake of sanity, I think we probably shouldn't mark them as "valid"

Reply:
Done.